dev/env.go

```diff
@@ -0,0 +1,240 @@
+	SetVars                                                []string
+}
+
+func (l ClusterHelmInstall) Init(
+	ctx context.Context, cluster *Cluster) error {
+	return cluster.HelmInstall(
+		ctx, cluster,
+		l.RepoName, l.RepoURL, l.PackageName, l.ReleaseName, l.Namespace, l.SetVars)
+}
+
+// Environment represents a development environment.
+type Environment struct {
+	Name string
+	// Working directory of the environment.
+	// Temporary files/kubeconfig etc. will be stored here.
+	WorkDir string
+	Cluster *Cluster
+	config  EnvironmentConfig
+}
+
+// Creates a new development environment.
+func NewEnvironment(name, workDir string, opts ...EnvironmentOption) *Environment {
+	env := &Environment{
+		Name:    name,
+		WorkDir: workDir,
+	}
+	for _, opt := range opts {
+		opt.ApplyToEnvironmentConfig(&env.config)
+	}
+	env.config.Default()
+	return env
+}
+
+// Initializes the environment and prepares it for use.
+func (env *Environment) Init(ctx context.Context) error {
+	kindConfig := `kind: Cluster
+apiVersion: kind.x-k8s.io/v1alpha4
+`
+
+	// Workaround for https://github.com/kubernetes-sigs/kind/issues/2411
+	// For BTRFS on LUKS.
+	if _, err := os.Lstat("/dev/dm-0"); err == nil {
+		kindConfig += `nodes:
+- role: control-plane
+  extraMounts:
+    - hostPath: /dev/dm-0
+      containerPath: /dev/dm-0
+      propagation: HostToContainer
+`
+	}
+
+	if err := os.MkdirAll(env.WorkDir, os.ModePerm); err != nil {
+		return fmt.Errorf("creating workdir: %w", err)
+	}
+
+	kindConfigPath := env.WorkDir + "/kind.yaml"
+	if err := ioutil.WriteFile(
+		kindConfigPath, []byte(kindConfig), os.ModePerm); err != nil {
+		return fmt.Errorf("creating kind cluster config: %w", err)
+	}
+
+	// Needs cluster creation?
+	var checkOutput bytes.Buffer
+	if err := env.execKindCommand(ctx, &checkOutput, nil, "get", "clusters"); err != nil {
+		return fmt.Errorf("getting existing kind clusters: %w", err)
+	}
+
+	// Only create cluster if it is not already there.
+	createCluster := !strings.Contains(checkOutput.String(), env.Name+"\n")
+
+	if createCluster {
+		// Create cluster
+		if err := env.execKindCommand(
+			ctx, os.Stdout, os.Stderr,
+			"create", "cluster",
+			"--kubeconfig="+env.Cluster.Kubeconfig, "--name="+env.Name,
+			"--config="+kindConfigPath,
+		); err != nil {
+			return fmt.Errorf("creating kind cluster: %w", err)
+		}
+	}
+
+	// Create _all_ the clients
+	cluster, err := NewCluster(
+		env.WorkDir+"/kubeconfig.yaml", env.config.ClusterOptions...)
+	if err != nil {
+		return fmt.Errorf("creating k8s clients: %w", err)
+	}
+	env.Cluster = cluster
+
+	// Run ClusterInitializers
+	if createCluster {
+		for _, initializer := range env.config.ClusterInitializers {
+			if err := initializer.Init(ctx, cluster); err != nil {
+				return fmt.Errorf("running cluster initializer: %w", err)
+			}
+		}
+	}
+
+	return nil
+}
+
+// Destroy/Teardown the development environment.
+func (env *Environment) Destroy(ctx context.Context) error {
```

Comment:
Should we check if the cluster exists before deletion?

Reply:
`kind delete cluster` does not fail when the cluster does not exist, so there is no need to check beforehand.
I also don't even think it's useful to return an "Cluster does not exist/Environment is already destroyed" error here, just because I can't think of any use case...